reps/2023-05-10-actors-batch-remote-api.md

```diff
@@ -0,0 +1,205 @@
+## Summary
+### General Motivation
+**Core Motivation**:
+1. Improve the performance of batch calling ActorTask. The current performance bottleneck of batch invoking ActorTask mainly lies in the duplicate serialization of parameters, repeated puts parameters into object store, and frequent context switching between Python and C++. The Batch remote API will optimize performance for these three aspects.
+
+
+In distributed computing scenarios, such as big data computing、AI training and inference, there is a need to send a large number of RPC requests(ActorTask) to multiple Actors in batches.  
+For example, in a typical star-topology architecture with one master Actor and 400 worker Actors, the same computational request needs to be sent to 400 worker Actors. 
+However, the computing tasks of each Worker Actor are very short. In such a scenario, the performance requirements for executing batch actor task remote of a large number of Actor are very high.  
+Therefore, for the scenario of batch calling Actor tasks, I want to add a new optimization API, batch_remote(), to improve the performance of batch submission of Actor Task calls.   
+After my own performance testing and comparison, this API can achieve a performance improvement of 40% ~ 98%.
+
+Current situation of batch calling actor tasks:
+```
+actors = [WorkerActor.remote() for _ in range(400)]
+
+# This loop's repeated invocation actually wastes a lot of performance.
+for actor in actors:
+  actor.compute.remote(args)
+```
+
+Using the new Batch Remote API:
+```
+actors = [WorkerActor.remote() for _ in range(400)]
+
+# Calling it only once can greatly improve performance.
+ray.experimental.batch_remote(actors).compute.remote(args)
+```
+
+The current performance bottleneck of batch invoking ActorTask mainly lies in the duplicate serialization of parameters, repeated puts parameters into object store, and frequent context switching between Python and C++.  
```

Comment:
Currently one way to avoid putting the same args repeatedly to object store is calling 'ray.put(args)' first and pass the obj ref to '.remote()'. Will this be good enough?

Reply:
The method you mentioned can only reduce the performance cost of repeatedly serializing parameters. However, the performance cost of frequent switching context between python and c++ cannot be avoided.   
You can see my performance test even without parameters. Batch remote API performance optimization can reach 40%.

![image](https://github.com/ray-project/enhancements/assets/11072802/2e284b31-4c5a-4ca0-8f8c-d361e46f5f1c)

Reply:
@jjyao Yes, so I'd like to introduce it as an internal API.